File: cracking_the_coding_interview_6th/chapter07/python/07_02_v02.py
from __future__ import annotations
from enum import Enum


class Rank(Enum):
    responder = 0
    manager = 1
    director = 2


class Caller:
    pass


class Employee:
    def __init__(self, handler: CallHandler):
        self.__current_call = None
        self._rank

    def recieve_call(self, call: Call):
        pass

    def call_completed(self):
        pass

    def escalate_and_reassign(self):
        pass

    def assign_new_call(self):
        pass

    def is_free(self) -> bool:
        return self.__current_call is not None

    @property
    def rank(self) -> Rank:
        return self._rank


class Director(Employee):
    def __init__(self, handler):
        super().__init__(handler)
        self._rank = Rank.director


class Manager(Employee):
    def __init__(self, handler):
        super().__init__(handler)
        self._rank = Rank.manager


class Responder(Employee):
    def __init__(self, handler):
        super().__init__(handler)
        self._rank = Rank.responder


class Call:
    def __init__(self):
        self.rank = Rank.responder
        self.__caller = None
        self.handler = None

    def reply(self, message: str):
        pass

    def incrementRank() -> Rank:
        pass

    def disconnect() -> None:
        pass


class CallHandler:
    __LEVELS = 3

    __NUM_RESPONDENTS = 10
    __NUM_MANAGERS = 4
    __NUM_DIRECTORS = 3

    def get_handler_for_call(self, call: Call) -> Employee:
        pass

    def dispatch_call_with_caller(self, caller: Caller) -> None:
        call = Call(caller)
        self.dispatch_call(call)

    def dispatch_call(self, call: Call) -> None:
        emp = self.get_handler_for_call(call)
        if emp:
            emp.recieve_call(call)
            call.handler = emp
        else:
            call.reply("Please wait for free employee to reply")
            self.call_queue[call.rank.value].append(call)

File: cracking_the_coding_interview_6th/chapter07/cpp/07_02_v02.cpp
// Cracking the Coding Interview p.308
#include <list>
#include <memory>
#include <string>
#include <vector>
using namespace std;
using ll = long long;

enum class Rank { responder = 0, manager, director };

class Employee;
class Call;
class CallHandler;

using EmployeePtr = shared_ptr<Employee>;
using CallPtr = shared_ptr<Call>;
using CallerPtr = shared_ptr<Caller>;

class Caller {};

class Employee {
  CallPtr _current_call = nullptr;

protected:
  Rank _rank;

public:
  Employee(CallHandler handler);
  void recieve_call(Call call);
  void call_completed();
  void escalate_and_reassign();
  bool assign_new_call();
  bool is_free() {
    return _current_call != nullptr;
  }
  Rank rank() {
    return _rank;
  }
};

class Director : public Employee {
public:
  Director(CallHandler handler) : Employee(handler) {
    _rank = Rank::director;
  }
};

class Manager : public Employee {
public:
  Manager(CallHandler handler) : Employee(handler) {
    _rank = Rank::manager;
  }
};

class Responder : public Employee {
public:
  Responder(CallHandler handler) : Employee(handler) {
    _rank = Rank::responder;
  }
};

class Call {
  Rank _rank;
  CallerPtr _caller;
  EmployeePtr _handler;

public:
  Call(CallerPtr c) : _rank(Rank::responder), _caller(c) {}

  void set_handler(EmployeePtr e) {
    _handler = e;
  }

  void reply(string message);
  Rank rank() {
    return _rank;
  }
  void set_rank(Rank r) {
    _rank = r;
  }
  Rank incrementRank();
  void disconnect();
};

class CallHandler {
  const ll LEVELS = 3;

  const ll NUM_RESPONDENTS = 10;
  const ll NUM_MANAGERS = 4;
  const ll NUM_DIRECTORS = 3;

  vector<vector<EmployeePtr>> employee_levels;
  vector<list<Call>> call_queues;

public:
  CallHandler(){};

  EmployeePtr get_handler_for_call(Call call);

  void dispatch_call(Caller caller) {
    Call call{make_shared<Caller>(caller)};
    dispatch_call(call);
  }

  void dispatch_call(Call call) {
    EmployeePtr emp = get_handler_for_call(call);
    if (emp) {
      emp->recieve_call(call);
      call.set_handler(emp);
    } else {
      call.reply("Please wait for free employee to reply");
      call_queues[static_cast<ll>(call.rank())].push_back(call);
    }
  }

  bool assign_and_call(Employee emp);
};

int main() {}